### ISSUE TITLE
Exact Matchers on get strings with '?' inside an AnyOf matcher return 501 since 0.18.0

### ISSUE THREAD

**Issue by thomasetter** [lipanski/mockito#86]
I have created a minimal [repro](https://github.com/thomasetter/mockito_query_bug_repro).

Bisecting pointed to #77 as the PR which broke this.

#83 could be related, as the symptoms in both cases are `501` errors.

The [test](https://github.com/thomasetter/mockito_query_bug_repro/blob/master/tests/repro.rs) which fails is:
```
#[test]
fn query_in_exact_test() {
    env_logger::init();
    let mocker = mockito::mock(
        "GET",
        Matcher::AnyOf(vec![
            Matcher::Exact("/bar?stuff".to_string()),
        ]),
    )
    .expect(1)
    .create();
    let moved_mocker = mocker;
    let client = reqwest::Client::new();
    println!(
        "{:?}",
        client.get(&format!("{}/bar?stuff", mockito::server_url())).send()
    );
    moved_mocker.assert();
}
```

I did not see any obvious reasons why this got broken by #77, do you have any ideas?

**Comment by lipanski** [comment]
@thomasetter I managed to figure out the problem. before #77 the path and query parts were being matched together. though #77  preserved backwards compatibility in regards to matching the query part in calls like `mock("GET", "/path?with_query")`, under the hood this is translated right away into something like `mock("GET", "/path").match_query("with_query")`. this happens here: https://github.com/lipanski/mockito/blob/5c7b6c172510a9378421a3ec0d1a350c602adcfc/src/lib.rs#L689-L700

the problem with that piece of code is that your `AnyOf` gets matched to the second pattern. `Mock#path` ends up getting set to `/bar?stuff` while `Mock#query` is set to `Matcher::Any`. because https://github.com/lipanski/mockito/pull/77/files#diff-4ce93534efc34e923ce01e975eb7ed80R18 is now splitting the path matching from the query matching, Mockito is trying to match the requested path "/bar" to the mocked path "/bar?stuff" and fails in the process.

a good workaround could be doing something like this:

```rust
let mocker = mockito::mock(
    "GET",
    Matcher::AnyOf(vec![
        Matcher::Exact("/bar".to_string()),
    ]),
)
.match_query("stuff")
.expect(1)
.create();
```

fixing this inside the crate could be done in two ways, but both aren't exactly ideal:

(1) that [piece of code](https://github.com/lipanski/mockito/blob/5c7b6c172510a9378421a3ec0d1a350c602adcfc/src/lib.rs#L689-L700) inside the constructor could traverse the matcher recursively in search of the query. however in the case of an `AnyOf` matcher it should consider the query part also an `AnyOf`, which in turn translate to something different than you intended in your initial code. this is what I mean:

```rust
# This accepts requests to either /bar?stuff or /foo?other
let _m1 = mockito::mock(
    "GET",
    Matcher::AnyOf(vec![
        Matcher::Exact("/bar?stuff".to_string()),
        Matcher::Exact("/foo?other".to_string()),
    ]),
)
.create();

# This accepts requests to /bar?stuff, /bar?other, /foo?stuff, /foo?other
let _m2 = mockito::mock(
    "GET",
    Matcher::AnyOf(vec![
        Matcher::Exact("/bar".to_string()),
        Matcher::Exact("/foo".to_string())
    ])
).match_query(Matcher::AnyOf(
    vec![
        Matcher::Exact("stuff"),
        Matcher::Exact("other")
])
.create();
```

(2) bring path and query matching back together, though this means breaking the interface again and being less flexible.

all in all, I'd be tempted to simply introduce a warning or an error message in these cases.

btw could you share your extended use case? how does your entire mock look like? is it going in the direction of my example above with `/foo?stuff` and `/bar?other` or can you use the workaround I suggested?

**Comment by thomasetter** [comment]
Thanks for the feedback!

I have looked into this some more and I think I found a way which should not break much.

I created a change in #87 which basically implements (2) but still supports the `match_query` interface.

What do you think, would something like that PR work for you?

**Comment by lipanski** [comment]
Released in 0.22.0
